nodejs/lib/util/http.js

```diff
@@ -180,16 +182,49 @@ var Http = (function() {
 		* (Ably and perhaps an auth server), so for efficiency, use the
 		* foreverAgent to keep the TCP stream alive between requests where possible */
 		var agentOptions = (rest && rest.options.restAgentOptions) || Defaults.restAgentOptions;
-		var doOptions = {headers: headers, encoding: null, agentOptions: agentOptions};
+		var doOptions = { headers: headers || undefined, responseType: 'buffer' };
 		if (body) {
 			doOptions.body = body;
 		}
 		if(params)
-			doOptions.qs = params;
+			doOptions.searchParams = params;
 
-		doOptions.uri = uri;
-		doOptions.timeout = (rest && rest.options.timeouts || Defaults.TIMEOUTS).httpRequestTimeout;
-		request[method](doOptions, handler(uri, params, callback));
+		if (!Http.agent) {
+			Http.agent = {
+				http: new http.Agent(agentOptions),
+				https: new https.Agent(agentOptions)
+			}
+		}
+
+		doOptions.agent = Http.agent;
+
+		let url;
+		if(uri.indexOf('?') > -1) {
```

Comment:
Ah, fair -- looks like I probably added that logic for authMethod=GET only just because if authMethod=POST, the `authParams` are sent in the post body not the query string, so no need to merge anything, so doing it stricter wasn't necessary to actually make it behave (as `request` was apparently less strict than `got` about query params being in the right place).

So should be fine to parse the queryParams in both cases, and then only mix in the authParams in the GET case

Reply:
@SimonWoolf I ran into some problems with this approach because the `type` param won't work if you send it in the body for an `authUrl` request so my solution was to mixin the provided querystring params into the body but also send them as part of the request search params (be95778). Let me know if you reckon that's okay.

Reply:
Sorry for the delay on this. What you've got looks fine.

(I'm a bit confused by the "my solution was to mixin the provided querystring params into the body but also.." statement because unless I'm missing something your code isn't doing that, it's just sending them as params. Which is good (mixing them into the body would be kindof weird)).